### ISSUE TITLE
[🐛 Bug]: CJS require of @wdio/cli and @wdio/reporter not implemented

### ISSUE THREAD

**Issue by vbyjsue** [webdriverio/webdriverio#9869]
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/webdriverio/webdriverio/blob/main/CONTRIBUTING.md#reporting-new-issues).

### WebdriverIO Version

8.4

### Node.js Version

16.13.2

### Mode

WDIO Testrunner

### What happened?

Tests are written in a cjs project with the testrunner started programmatically. When upgrading to v8 some wdio modules are no longer possible to require. 
```
const Launcher = require('@wdio/cli').default
const WDIOReporter = require('@wdio/reporter').default;
```
fails since they only support esm import.
A temporary workaround is to use ```const { Launcher } = await import('@wdio/cli');```
but in the case of @wdio/reporter that is a bit trickier.

### What is your expected behavior?

To be able to use 
```
const Launcher = require('@wdio/cli').default
const WDIOReporter = require('@wdio/reporter').default;
```

### How to reproduce the bug.

Use 
```
require('@wdio/cli').default
```
or
```
require('@wdio/reporter').default
```
in v8

### Relevant log output

```typescript
Launcher failed to start the test Error [ERR_REQUIRE_ESM]: require() of ES Module
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### Is there an existing issue for this?

- [X] I have searched the existing issues

**Comment by christian-bromann** [comment]
Thanks for reporting!

I would suggest to for the `@wdio/cli` package to do the same approach as we have for others which is create a cjs directory where we re-export the same interface in CJS by dynamically importing the ESM module.

For `@wdio/reporter` I don't see a reason why it should be ported back to CJS given that you only need to import it for custom reporter which can easily be transformed to ESM without having to transform all the other code.

Any contributions that resolves the bug are highly appreciated. Please take a look into our [contribution guidelines](https://webdriver.io/docs/contribute) and let us know if you have any questions. Cheers!

**Comment by olivier-martin-sf** [comment]
@christian-bromann I found the same apply to the `@wdio/logger`. We are using the `@wdio/logger` as part of a custom service. Do you recommend to wrap the `@wdio/logger` in a similar fashion than `@wdio/cli`? I think the main reason why it failed it's because of the top-level await here: https://github.com/webdriverio/webdriverio/blob/c58439ce91910803b53b0337ac01e3b973350925/packages/wdio-logger/src/index.ts#L11

**Comment by christian-bromann** [comment]
The logger would need to have a separate `tsconfig.cjs.json` that compiles the whole package as CJS artifact. This will be difficult though because we use a top level await in the package. Using a CJS fork or the v7 package would not help because you would end up having two logger packages trying to write to the same files and overwrite each others content. We really recommend to upgrade your tests (not whole project, only e2e tests) to ESM.

**Comment by olivier-martin-sf** [comment]
Thank you! So it would mean that the consumer of our service would have to opt-in their tests to ESM?

**Comment by christian-bromann** [comment]
Users should be able to run CJS tests with WebdriverIO v8 even if all WebdriverIO service packages are written in ESM.

**Comment by olivier-martin-sf** [comment]
I've tried to bump the WebdriverIO dependencies to v8 in the service that I am working on (the only direct dependency that we have is the `@wdio/logger`). Our WebdriverIO service is written in TypeScript and compiled down as CJS as most of our consumers codebases are still using CJS. 

I've then try to use that service in a repo that uses WebdriverIO v8 & CJS and it fails:

```
2023-04-05T17:10:25.119Z ERROR @wdio/config:ConfigParser: Failed loading configuration file: file:///Users/john.doe/project/packages/integration-wdio/wdio.conf.js: require() of ES Module /Users/john.doe/project/node_modules/@wdio/logger/build/index.js from /Users/john.doe/project/packages/wdio-project-service/build/adapters/mobile-driver-adapter.js not supported.
Instead change the require of index.js in /Users/john.doe/project/packages/wdio-project-service/build/adapters/mobile-driver-adapter.js to a dynamic import() which is available in all CommonJS modules.
```

That's how TypeScript is compiling the logger import and what's causing the issue:

```
const logger_1 = __importDefault(require("@wdio/logger"));
```

Am I missing something? I am talking from the perspective of a custom service author that push a package on npm

**Comment by christian-bromann** [comment]
> Our WebdriverIO service is written in TypeScript and compiled down as CJS as most of our consumers codebases are still using CJS.

I recommend have the service create artifacts for ESM and CJS so it can be potentially used for v7 CJS users as well as v8 ESM ones. Have a close look how some of our [other service](https://github.com/webdriverio-community/wdio-edgedriver-service/blob/main/src/cjs/index.ts) do this. You basically write your service in ESM and have a CJS file that imports the ESM package and just re-exports it.

**Comment by christian-bromann** [comment]
Keeping this open as we still should provide a CJS artifact for `@wdio/reporter`

**Comment by christian-bromann** [comment]
Actually thinking about this another time, it doens't make sense to have a CJS artifact for `@wdio/reporter`. If you want to build a reporter that works for both `v7` and `v8` and uses CJS you should build it on top of the last `v7` version of `@wdio/reporter`. Unless there are any other reasons why a CJS version of this package is required I am gonna go close this.

Thanks @DirkoOdendaal for providing a patch for the `@wdio/cli` package.

**Comment by ashil88** [comment]
Hi

Just wondering if any additional changes needed to be made in order to get this to work post the fix? I'm using v8.10.2, but

```
const Launcher = require('@wdio/cli').default;
...
const wdio = new Launcher( ... );
```

throws a `TypeError: Launcher is not a constructor` error for me.

Thanks
Ashil

**Comment by DirkoOdendaal** [comment]
Hi @ashil88 ,

I found a bug that I'm working on fixing, but I doubt that it's the error you are getting.
Do you have a sample project where I can see the error?

You could also try:
`const {Launcher} = require('@wdio/cli');`

Don't believe that @wdio/cli has a default export and it only exposes named exports.

**Comment by vbyjsue** [comment]
And you have tried the proposed workaround?
`const { Launcher } = await import('@wdio/cli');`

**Comment by ashil88** [comment]
Hi @DirkoOdendaal 

Thanks, `const {Launcher} = require('@wdio/cli');` seems to have moved things on a bit. I can create a new instance of Launcher, but now getting a error with `run`. In `\node_modules\@wdio\cli\build\cjs\index.js` it seems that `this.#esmLauncher` isn't defined for me in the `run()` function; possibly there's an issue with the import on line 11. Will keep trying things and report back if I find a solution.

Thanks
Ashil

**Comment by DirkoOdendaal** [comment]
Hi @ashil88 ,
That is exactly the bug I'm working on.
This is a timing issue and as a workaround you can add a timeout just before the `run()`.
The issue is that the constructor does an async call, but the constructor can't be awaited. So working on moving the async call for now.

Hope that helps you!

**Comment by ShradhaSangtani** [comment]
Hello @DirkoOdendaal , 
The issue is not fixed yet. The following line of code doesn't launch the runner because of null value for 'esmLauncher'.
https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-cli/src/cjs/index.ts#LL20C8-L20C8
As a workaround dynamic import for '@wdio/cli' worked.
`const { Launcher } = await import('@wdio/cli');`

But the issue with the CJS module is not fixed. Can you please give us an update on this ? 
Thanks.

**Comment by christian-bromann** [comment]
A new PR with a fix was raised in #10607

**Comment by oldmankit** [comment]
I'm on webdriverio version 8.13.1 and seeing this problem with requiring `wdio-video-reporter`. It seems like this should already be fixed by that latest release, unless I'm missing something?

**Comment by christian-bromann** [comment]
The `@wdio/reporter` package is still ESM only.

**Comment by ForesightImaging** [comment]
In 7.x I was able to write my config file in TS and import my custom reporter written in TS all using the same options as my tests written in TS thanks ts-node and smart engineering. Now this 'internal' decision breaks two of those at a time. I see authors using [sneaky tricks](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-junit-reporter/src/index.ts) like importing .js files in TS to avoid proper imports. advocating top-level awaits or changing entire project module types in package.js to get things going again. This is a community friction issue that isn't going to go away. 8.0 has been out for a year and you're getting lots of issues.

**Comment by christian-bromann** [comment]
This is an odd comment but I am happy to answer to it.

> I see authors using [sneaky tricks](https://github.com/webdriverio/webdriverio/blob/main/packages/wdio-junit-reporter/src/index.ts) like importing .js files in TS to avoid proper imports.

We have spend a lot of time to update the code base to ESM to ensure we can update dependencies in the future and don't risk any security vulnerabilities through transient dependencies. In ESM you have to provide a file extension to import modules and we import JS files because that is what the reality looks like after the code base is transformed. That is not a sneaky trick at all and used in many modern JavaScript codebases.

> advocating top-level awaits

Not sure where this project advocates this, also please provide information why you disagree with top level awaits.

> changing entire project module types in package.js to get things going again.

That is totally unnecessary as WebdriverIO is very well able to run in CJS context. As long as your project is set up correctly you should have no problem running tests in CJS using `require`.

> This is a community friction issue that isn't going to go away. 8.0 has been out for a year and you're getting lots of issues.

Many issues around ESM/CJS compatibility have been resolved short after the v8 release. I haven't seen any issues coming in for some time now. So not sure where this comes from.

I appreciated your feedback and can understand any frustration that might have come up. However going into an Open Source project and releasing ones own frustration to the community is not a good approach. If you like to see changes or improve the compatibility with WebdriverIO, join the community and raise pull requests with improvements.

**Comment by oldmankit** [comment]
> The `@wdio/reporter` package is still ESM only.

Sorry that I don't understand this -  I'm not sure what it means that this issue is fixed, but is "ESM only".

This is just my ignorance, but at the moment I can't use the 'wdio-video-reporter' package, as with

`const video = require('wdio-video-reporter');`

The output includes `ERROR @wdio/config:ConfigParser: Failed loading configuration file: file:///folder/wdio.conf.js: require() of ES Module folder\node_modules\@wdio\reporter\build\index.js from folder\node_modules\wdio-video-reporter\dist\wdio-video-reporter.cjs not supported.
Error [ERR_REQUIRE_ESM]: require() of ES Module `

This looked like the same problem reported in this issue.

Am I waiting for wdio-video-reporter to be fixed too? Sorry if this is all obvious or if I should be creating a new issue.

**Comment by christian-bromann** [comment]
Which version of the video reporter are you using? You should be able to use the reporter without having it to import it via:

```js
 reporters: [
    ['video', {
      saveAllVideos: false,       // If true, also saves videos for successful test cases
      videoSlowdownMultiplier: 3, // Higher to get slower videos, lower for faster videos [Value 1-100]
    }],
  ],
```

**Comment by oldmankit** [comment]
I'm using wdio-video-reporter version 4.0.3.

> You should be able to use the reporter without having it to import it. 

Yes, that did work, thank you! (I wonder if the [documentation for the video reporter](https://webdriver.io/docs/wdio-video-reporter/#add-the-reporter-to-config) needs to be updated, as it uses the require statement.)

**Comment by christian-bromann** [comment]
> (I wonder if the [documentation for the video reporter](https://webdriver.io/docs/wdio-video-reporter/#add-the-reporter-to-config) needs to be updated, as it uses the require statement.)

Yes, @oldmankit would you mind raising a PR for this?

**Comment by christian-bromann** [comment]
Raised a PR for adding CJS support for the reporter package so your custom reporter can stay CJS modules: https://github.com/webdriverio/webdriverio/pull/10998

**Comment by rsinghi** [comment]
Is this issue resolved? I am using webdriverio@8.15.6 (where the PR was merged), but still seeing the issue: 
Code: 
const WDIOReporter = require('@wdio/reporter');

Error:
rror [ERR_REQUIRE_ESM]: require() of ES Module node_modules/@wdio/reporter/cjs/index.js is not supported.
index.js is treated as an ES module file as it is a .js file whose nearest parent package.json contains "type": "module" which declares all .js files in that package scope as ES modules.
Instead rename index.js to end in .cjs, change the requiring code to use dynamic import() which is available in all CommonJS modules, or change "type": "module" to "type": "commonjs" in node_modules/@wdio/reporter/package.json to treat all .js files as CommonJS (using .mjs for all ES modules instead).

**Comment by christian-bromann** [comment]
see #11001
